src/openforms/forms/migrations/0029_make_logic_order_explicit.py

```diff
@@ -0,0 +1,126 @@
+# Generated by Django 3.2.13 on 2022-06-23 11:02
+
+from collections import defaultdict
+from itertools import chain
+
+from django.db import migrations
+
+from openforms.formio.utils import iter_components
+from openforms.utils.json_logic import JsonLogicTest
+
+
+def find_component_key(test: JsonLogicTest, candidates: list) -> str:
+    for value in test.values:
+        if isinstance(value, JsonLogicTest) and value.operator == "var":
+            candidate = value.values[0]
+            if candidate in candidates:
+                return candidate
+
+        elif isinstance(value, JsonLogicTest) and value.operator == "date":
+            return find_component_key(value, candidates=candidates)
+
+    return ""
+
+
+def make_logic_order_explicit(apps, schema_editor):
+    """
+    Set an explicit order value for logic rules, per form.
+
+    Since there was no explicit ordering beforehand, we apply the following logic:
+
+    1. Simple rules: introspect the trigger and check which key it is for
+        1. Look up the key in the form components
+        2. Derive which form step the key belongs to
+        3. Order based on form step order
+        4. Within a form step, order on component hierarchy, depth first (i.e. we fetch
+           child components before stepping to siblings.
+        5. If still ambiguous, order by PK
+    2. Simple rules that failed introspection
+    3. Advanced rules: order by PK
```

Comment:
Ewen remarked that the `find_component_key` construct is smart enough to be able to deeply inspect advanced logic rules as well, so we could leverage this to improve the ordering of those. To decide if we want to do this or not!

Reply:
Taking a look at this now